straxen/scada.py

```diff
@@ -0,0 +1,246 @@
+                     start=None,
+                     end=None,
+                     context=None,
+                     run_id=None,
+                     time_selection_kwargs={'full_range': True},
+                     value_every_seconds=1):
+    """
+    Function which returns XENONnT slow control values for a given set
+    of parameters and time range.
+
+    The time range can be either defined by a start and end time or via
+    the run_id, target and context.
+
+    :param parameters: dictionary containing the names of the requested
+        scada-parameters. The keys are used as identifier of the
+        parameters in the returned pandas.DataFrame.
+    :param start: int representing the start time of the interval in ns
+        unix time.
+    :param end: same as start but as end.
+    :param context: Context you are working with (e.g. st).
+    :param run_id: Id of the run. Can also be specified as a list or
+        tuble of two run ids. In this case we will return the time
+        range lasting between the start of the first and endtime of the
+        second run.
+    :param time_selection_kwargs: Keyword arguments taken by
+        st.to_absolute_time_range(). Default: full_range=True.
+    :param value_every_seconds: Defines with which time difference
+        values should be returned. Must be an integer!
+        Default: one value per 1 seconds.
+    :return: pandas.DataFrame containing the data of the specified
+        parameters.
+    """
+    if not isinstance(parameters, dict):
+        mes = 'The argument "parameters" has to be specified as a dict.'
+        raise ValueError(mes)
+
+    if np.all((run_id, context)):
+        # User specified a valid context and run_id, so get the start
+        # and end time for our query:
+        if isinstance(run_id, (list, tuple)):
+            start, _ = context.to_absolute_time_range(run_id[0], **time_selection_kwargs)
+            _, end = context.to_absolute_time_range(run_id[-1], **time_selection_kwargs)
+        else:
+            start, end = context.to_absolute_time_range(run_id, **time_selection_kwargs)
+
+    if not np.all((start, end)):
+        # User has not specified any vaild start and end time
+        mes = ('You have to specify either a run_id and context.'
+               ' E.g. call get_scada_values(parameters, run_id=run,'
+               ' target=raw_records", context=st) or you have to specifiy'
+               'a valid start and end time.')
+        raise ValueError(mes)
+
+    # Now loop over specified parameters and get the values for those.
+    for ind, (k, p) in enumerate(parameters.items()):
+        print(f'Start to query {k}: {p}')
+        temp_df = _query_single_parameter(start, end, k, p, value_every_seconds)
+
+        if ind:
+            m = np.all(df.loc[:, 'time'] == temp_df.loc[:, 'time'])
+            mes = ('This is odd somehow the time stamps for the query of'
+                   f' {p} does not match the other time stamps.')
```

Comment:
This can happen right if the loop takes long and you specify a time in the future?